src/utils/include/c_resource.hpp

```diff
@@ -0,0 +1,366 @@
+    {
+        if (p != initValue)
+            destructFunction(&p);
+    }
+
+public:
+    // default constructor for API schema 1
+    [[nodiscard]] constexpr c_resource() noexcept
+        requires requires { constructFunction(); } && std::is_invocable_r_v<T, TConstructor>
+    {
+        _element = constructFunction();
+    }
+
+    // default constructor for API schema 2
+    [[nodiscard]] constexpr explicit c_resource() noexcept
+        requires requires { construct(&_element); }
+        : _element{ initValue } {
+        construct(&_element);
+    }
+
+    [[nodiscard]] constexpr explicit c_resource(construct_t) noexcept
+        requires std::is_invocable_r_v<T, TConstructor>
+        : _element{ constructFunction() } {}
+
+    template<typename... Ts>
+        requires(std::is_invocable_r_v<TConstructReturn, TConstructor, Ts...> || std::is_invocable_r_v<TConstructReturn, TConstructor, Ts..., bool>)
+    [[nodiscard]] constexpr explicit(sizeof...(Ts) == 1) c_resource(Ts &&...Args) noexcept {
+        _element = [](auto &&...args) -> decltype(auto) { // construct wrapper to allow for default constructor arguments
+            return constructFunction(std::forward<decltype(args)>(args)...);
+        }(std::forward<Ts>(Args)...);
+    }
+
+    template<typename... Ts>
+        requires(std::is_invocable_v<TConstructor, T *, Ts...>)
+    [[nodiscard]] constexpr explicit(sizeof...(Ts) == 1) c_resource(Ts &&...Args) noexcept {
+        if constexpr (std::is_pointer_v<T>) {
+            _element = initValue;
+            constructFunction(&_element, std::forward<Ts>(Args)...);
+        } else {
+            _element = [](auto &&...args) -> decltype(auto) { // construct wrapper to allow for default constructor arguments
+                return constructFunction(std::forward<decltype(args)>(args)...);
+            }(std::forward<Ts>(Args)...);
+        }
+    }
+
+    template<typename... Ts>
+        requires(std::is_invocable_v<TConstructor, T *, Ts...>)
+    [[nodiscard]] constexpr auto emplace(Ts &&...Args) noexcept
+        requires std::is_pointer_v<T>
+    {
+        destruct(_element);
+        _element = initValue;
+        return construct(&_element, static_cast<Ts &&>(Args)...);
+    }
+
+    [[nodiscard]] constexpr c_resource(c_resource &&other) noexcept
+        requires std::is_pointer_v<T>
+    {
+        _element       = other._element;
+        other._element = initValue;
+    };
+    constexpr c_resource &operator=(c_resource &&rhs) noexcept
+        requires std::is_pointer_v<T>
+    {
+        if (this != &rhs) {
+            destruct(_element);
+            _element     = rhs._element;
+            rhs._element = initValue;
+        }
+        return *this;
+    };
+    constexpr void swap(c_resource &other) noexcept {
+        auto ptr       = _element;
+        _element       = other._element;
+        other._element = ptr;
+    }
+
+    constexpr ~c_resource() noexcept = delete;
```

Comment:
`constexpr ~c_resource() noexcept = delete;`

should be

`constexpr ~c_resource() noexcept {}`

Also, destructors are implicitly `noexcept(true)` IIRC

Reply:
The default constructor is explicitly deleted in favour of the following specialised versions.